### ISSUE TITLE
SFTP Initial setup, then send basic file from Azure blob storage to ReportStream

### ISSUE THREAD

**Issue by scleary1cs** [CDCgov/trusted-intermediary#1075]
# Story
As a PHL, in order for my results to be sent to RS, I need to send an ORU result from an SFTP site.

## Acceptance Criteria
- [x] A file can be read from Azure blob storage
- [x] A file's contents can be sent to RS
- [x] Uses the flexion org as sender (for now)
- [x] Quality scanning by sonar cloud
- [x] Automated security vulnerability updates

## Non - Acceptance Criteria
- [X] Doesn't need to be automatically triggers
- [X] Doesn't need to handle zip files
- [X] Doesn't need to handle password protected files

### Research
- [ ] *Research work needed to complete the story*
- [ ] Foundational: *High-level research that will support this and future efforts*

###  Engineering
- [x] Github Setup
  - [x] Protect Main branch
  - [x] Link Repo to zenhub board
  - [x] Require Approvals on PRs
  - [x] Add .gitignore file
  - [x] Add code quality and security stuff to pipeline (in future we may want API testing, but we don't have APIs yet)
    - [x] codeql
    - [x] sonarcloud (Neil working on this - was waiting on access to CDC org)
    - [x] SCA (software composition analysis) - we use Snyk in TI
  - [x] Compile app upon merge into main
  - [x] Add deploy step for binary to pipeline (deploying to PR when terraform changes, deploying to staging on merge to main. Deploys available for dev and internal)
  - [x] Require commit signing
  
   

- [x] Repo Setup
  - [x] Add Dependabot
  - [x] Add issue templates
  - [x] Add updated  CI/CD files
  - [x] ADR for Using GoLang
  - [x] ADR for testify framework
  - [x] Copy skeleton to README.md
  - [x] Tidy up the default Readme files
  - [x] Pre-commit hooks (linting, secret scanning, etc)
  - [x] Enable SFTP for Azure Storage account 
  - [x] Terraform Setup: branch terraform
    - [x] Reach out to CDC to get a new vnet - @halprin 
    - [x] Swap temporary Vnet for official Vnet in Terraform.
    - [x] Create blob storage in all envs via terraform
    - [x] App Service
      - [x] Adjust tier for app service 
    - [x] Additional vnet
    - [x] App service placed inside of private vnet
    - [x] Deny All public traffic 
    - [x] Logs
    - [x] dev
    - [x] internal
    - [x] staging
    - (Not adding prod environment for now, we'll do it when we're ready)

- [x] Development
  - [x] Code to read a file and send to report stream
  - [x] Environment variable setup
  - [x] Write generic interfaces for Azure bucket and ReportStream sending.
  - [x] Make Azure-specific implementation that adheres to the above interface.
  - [x] Make a ReportStream-specific implementation that adheres to the above interface.
  - [x] Docker Compose for setting up Azurite with volume (so we can get at the files locally).
  - [x] Have docker-compose set up the container and a test file
  - [x] Auth with RS using JWT/token/login endpoint
    - [x] Generate JWT
    - [x] Extract JWT
    - [x] Call Auth endpoint with JWT
    - [x] Extract Bearer token for sending and call RS
      - [x] Get credentials from Azure Key Vault (needs TF setup) for existing Flexion fake public health lab
      - [X] Need to update vault name in `key.tf` to be 4 characters shorter so it works in `internal` env. 
      - [x] Will need to update staging key value again after creating new vault in staging TF
   - [x] Clean up all to-do(s) by doing them or creating new stories
## Definition of Done
- [x] Documentation tasks completed
  - [x] Documentation and diagrams created or updated
    - [x] Implementation guide (`/ig` folder)
    - [x] ADRs (`/adr` folder)
    - [x] Main `README.md`
    - [x] Other READMEs in the repo
    - [x] If applicable, update the `ReportStream Setup` section in `README.md`
  - [x] [Threat model](https://lucid.app/lucidchart/693ff5f9-c289-4585-9e74-a7f5abb36d69/edit) updated
  - [x] API documentation updated
- [x] Code quality tasks completed
  - [x] Code refactored for clarity and no design/technical debt
  - [x] Adhere to separation of concerns; code is not tightly coupled, especially to 3rd party dependencies
  - [x] Code is reviewed or developed by pair; 1 approval is needed but consider requiring an outside-the-pair reviewer
  - [x] Code quality checks passed
- [x] Security & Privacy tasks completed
  - [x] Security & privacy gates passed
- [x] Testing tasks completed
  - ~~Load tests passed~~
  - ~~Unit test coverage of our code >= 90%~~
- [x] Build & Deploy tasks completed
  - [x] Build process updated
  - ~~API(s) are versioned~~
  - [x] Feature toggles created and/or deleted.  Document the feature toggle
  - [x] Source code is merged to the main branch

## Research Questions
- *Optional: Any initial questions for research*

## Decisions
- *Optional: Any decisions we've made while working on this story*

## Notes
0. Have a service, do TF setup (need blob storage, maybe new resource group, app service or Azure functions, private vnet, ci/cd pipelines)(may need db in future, but will wait until that's more clearly defined)
1. read from Azure blob storage
2. Log into RS (using a shared organization/sender, and using a shared key - can probably use Flexion sender/receiver for initial testing).
3. Send file to existing RS endpoint

**Comment by halprin** [comment]
Azure help desk ticket `REQ0302028` created to track the creation of three new VNETs in our three Azure environments.

**Comment by scleary1cs** [comment]
Blocked on sonar cloud, and Vnet.

**Comment by halprin** [comment]
I've got a meeting on the books with Jeremy Page, who is an admin of the SonarCloud organization, to get us unblocked.
